Remove redundant tags (and actions)

marcstern opened the pull request:
This removes redundant tags like tag:xss,tag:bla,tag:xss.
Very useful when using mod_macro or Define, so the same tag may be added at several places.
Does the same for other actions like logdata.
Detects redundancy only if the content of the action is 100% the same.

Commit:
Check for invalid pointers

Commit:
https://github.com/SpiderLabs/ModSecurity/issues/533

Commit:
another NULL pointer check

Commit:
another NULL pointer check

Commit:
Approach version

Commit:
Add environment variable with mod_security2 version number

Commit:
Remove redundant tags (and actions)

martinhsv commented:
Hi @marcstern ,

Isn't this pretty similar to #2685 which we have discussed previously?

As before, my general leaning is that the ModSecurity engine should avoid second-guessing what the rule writers have done by doing autocorrections.

marcstern commented:
Hi @martinhsv, it's indeed a formal PR for what we discussed.
I know you were not in favor of this, but I'm proposing it anyway. In this case, it's really a simple clean-up that saves RAM and CPU, so I really think it has a value.
Also keep in mind that rule writers, in professional or hosted environments, don't have the full control like you have when working on a stand-alone version. You inherit (part of) rules that you can extend but you don't always have the full picture and, often, you combine different pieces that have to set a tag but can be combined together as well.

Commit:
https://github.com/SpiderLabs/ModSecurity/issues/2927

Commit:
Logging enhancement: show if rule is chained, in case of no id, show begin of rule

Commit:
Allow multiple digits in sanitizeMatchedBytes

Commit:
remove useless apr_pstrdup()

Commit:
remove useless memset

Commit:
Handle capture as tx.1=char in validateByteRange

Commit:
Centralized function compatible with Linux & Windows (also with mpm-itk & mod_ruid2) to get username

Commit:
Revert "Centralized function compatible with Linux & Windows (also with mpm-itk & mod_ruid2) to get username"

This reverts commit c7b28f0e13fa18d8503877e026744f76607cfbae.

Commit:
Centralized function to get user name, compatible with Linux & Windows (also with mpm-itk & mod_ruid2)

Commit:
Compatibility with libyajl decoding the buffer inline

Commit:
Fix ErrorDocument processing if status is changed during phase 4

Commit:
Update apache2_io.c

Another NULL pointer check

Commit:
Check return code of apr_procattr_io_set()

Commit:
Fixed 2 memory leaks

Commit:
removed useless code

Commit:
Fix for DEBUG_CONF compile flag

Commit:
Support for "filename*" in multipart (precedence over "filename").
Reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition

Commit:
Fix for https://github.com/SpiderLabs/ModSecurity/issues/610

Commit:
status 400 instead of 500 on parsing error

Commit:
cleanup

Commit:
Allow macro expansion in ctl:ruleRemoveTargetByTag

Commit:
Merge pull request #30 from marcstern/v2/approach/version

Approach version

Commit:
Merge pull request #31 from marcstern/v2/mst/expand_ruleRemoveTargetBy

Allow macro expansion in ctl:ruleRemoveTargetByTag

Commit:
Merge pull request #32 from marcstern/v2/mst/nullcheck

V2/mst/nullcheck

Commit:
Merge pull request #33 from marcstern/v2/mst/username

V2/mst/username

Commit:
Merge pull request #34 from marcstern/v2/mst/yajl_compat

V2/mst/yajl compat

Commit:
Merge pull request #35 from marcstern/v2/mst/issue533

https://github.com/SpiderLabs/ModSecurity/issues/533

Commit:
Merge branch 'v2/latest' into v2/mst/issue2849

Commit:
Merge pull request #36 from marcstern/v2/mst/issue2849

Fix ErrorDocument processing if status is changed during phase 4

Commit:
Merge pull request #37 from marcstern/v2/mst/json400

status 400 instead of 500 on parsing error

Commit:
Merge pull request #38 from marcstern/v2/mst/issue610

Fix for https://github.com/SpiderLabs/ModSecurity/issues/610

Commit:
Merge pull request #39 from marcstern/v2/mst/filenamestar

Support for "filename*" in multipart (precedence over "filename").

Commit:
Merge pull request #40 from marcstern/v2/mst/debug_conf

Fix for DEBUG_CONF compile flag

Commit:
Merge pull request #41 from marcstern/v2/mst/mem_leak1

V2/mst/mem leak1

Commit:
Merge pull request #42 from marcstern/v2/mst/rc_check

Check return code of apr_procattr_io_set()

Commit:
Merge pull request #43 from marcstern/v2/mst/capturebyte

Handle capture as tx.1=char in validateByteRange

Commit:
Merge pull request #44 from marcstern/v2/mst/optim2

remove useless memset

Commit:
Merge pull request #45 from marcstern/v2/mst/optim1

remove useless apr_pstrdup()

Commit:
Merge pull request #46 from marcstern/v2/mst/sanit_bytes

Allow multiple digits in sanitizeMatchedBytes

Commit:
Merge pull request #47 from marcstern/v2/mst/log_noid

Logging enhancement: show if rule is chained, in case of no id, show …

Commit:
Merge pull request #48 from marcstern/v2/mst/issue2927

https://github.com/SpiderLabs/ModSecurity/issues/2927

Commit:
Merge pull request #49 from marcstern/v2/mst/redundant_actions

Remove redundant tags (and actions)

Commit:
Merge pull request #50 from marcstern/v2/mst/version

Add environment variable with mod_security2 version number

Commit:
Merge pull request #51 from marcstern/v2/approach/version

Approach version

Commit:
set pointer to NULL after free

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update mod_security2.c

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
NULL pointer check

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Merge pull request #52 from marcstern/v2/mst/yajl_compat

Compatibility with libyajl decoding the buffer inline

Commit:
Merge branch '!latest' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update msc_json.c

Commit:
Merge branch 'v2/mst/yajl_compat' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update msc_json.c

Commit:
Merge branch 'v2/mst/yajl_compat' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update msc_logging.c

Commit:
Merge branch 'v2/mst/username' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update msc_logging.c

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update msc_reqbody.c

Commit:
Merge branch 'v2/mst/json400' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update persist_dbm.c

Commit:
Merge branch 'v2/mst/username' of https://github.com/marcstern/ModSecurity into !latest

Commit:
NULL pointer check

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update re_actions.c

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update re_operators.c

Commit:
Merge branch 'v2/mst/optim2' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update re_operators.c

Commit:
Merge branch 'v2/mst/optim2' of https://github.com/marcstern/ModSecurity into !latest

Commit:
; incorrectly replaced by space in cmdline

In cmdline, ; was incorrectly repalced by a space. As ; is a command separator, this leads to false negative (missing attacks).

Commit:
Merge pull request #53 from marcstern/v2/mst/cmdline-semicolon

; incorrectly replaced by space in cmdline

Commit:
Merge branch '!latest' of https://github.com/marcstern/ModSecurity into !latest

Commit:
typo

Commit:
Merge branch 'v2/mst/mem_leak1' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Double memory allocation:
'current_targets' is allocated in ruleset->mp.
'rule->p1' is a copy of current_targets, but we strdup it in the same memory pool as 'current_targets'.
So, simply assign 'current_targets' to 'rule->p1'.

Commit:
Merge pull request #54 from marcstern/v2/mst/optim3

Fixed double memory allocation

Commit:
Merge branch '!latest' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update re_operators.c

Commit:
other NULL checks

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Fixed NULL check

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Merge pull request #56 from marcstern/v2/mst/expand_ruleRemoveTargetBy

Update re_operators.c

Commit:
Fixed merge problem

Commit:
Another NULL check fix

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Another NULL check fix

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Another NULL check

Commit:
Update re.c

Commit:
Merge branch 'v2/mst/log_noid' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Update re.c

Commit:
Merge branch 'v2/mst/log_noid' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Another NULL check

Commit:
Another NULL check

Commit:
Fixed another NULL check

Commit:
Other NULL checks

Commit:
Update apache2_config.c

Commit:
Update apache2_config.c

Commit:
Merge branch 'v2/mst/log_noid' of https://github.com/marcstern/ModSecurity into !latest

Commit:
Merge branch 'v2/mst/nullcheck' of https://github.com/marcstern/ModSecurity into !latest

Commit:
PCRE2 support

Commit:
Merge branch '!latest' of https://github.com/marcstern/ModSecurity into v2/mst/redundant_actions

Commit:
Update re.c

added the "sanitizeXXX" variants

martinhsv commented:
Hello @marcstern ,

I'm going to go ahead and close this.

For the main proposal -- which you have agreed is identical to your earilier PR -- it is for the same reasons. While I understand the outcome of #2685 was not your own preference, I did articulate my reasons.

Note that the current version of this PR contains a lot of disparate changes unrelated to the stated purpose of the PR. PRs that are either very large and/or contain changes unrelated to each are problematic to manage.

PRs are more likely to be actively considered if they are both tightly focussed and result in a tangible improvement for the users.
